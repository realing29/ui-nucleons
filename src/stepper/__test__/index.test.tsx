import React from 'react';
import { mount } from 'enzyme';
import { act, Simulate } from 'react-dom/test-utils';
import { Stepper } from '..';

describe('<Stepper />', () => {
  it('should render without props', () => {
    const wrapper = mount(
      <Stepper />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle input props', () => {
    const focusSpy = jest.fn();
    const blurSpy = jest.fn();

    const wrapper = mount(
      <Stepper
        defaultValue={123}
        disabled
        onFocus={focusSpy}
        onBlur={blurSpy}
      />
    );

    expect(blurSpy).toBeCalledTimes(0);
    expect(focusSpy).toBeCalledTimes(0);

    act(() => {
      Simulate.focus(wrapper.find('[data-testid="stepper:input"]').getDOMNode());
    });
    wrapper.update();

    expect(focusSpy).toBeCalledTimes(1);
    expect(blurSpy).toBeCalledTimes(0);

    act(() => {
      Simulate.blur(wrapper.find('[data-testid="stepper:input"]').getDOMNode());
    });
    wrapper.update();

    expect(focusSpy).toBeCalledTimes(1);
    expect(blurSpy).toBeCalledTimes(1);
  });

  it('should handle "onFocus" missing', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <Stepper
        onFocus={spy}
      />
    );

    expect(spy).toBeCalledTimes(0);

    act(() => {
      Simulate.focus(wrapper.find('[data-testid="stepper:input"]').getDOMNode());
    });
    wrapper.update();
    expect(spy).toBeCalledTimes(1);

    wrapper.setProps({ onFocus: undefined });

    act(() => {
      Simulate.focus(wrapper.find('[data-testid="stepper:input"]').getDOMNode());
    });
    wrapper.update();
    expect(spy).toBeCalledTimes(1);
  });

  it('should handle "onBlur" missing', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <Stepper
        onBlur={spy}
      />
    );

    expect(spy).toBeCalledTimes(0);

    act(() => {
      Simulate.blur(wrapper.find('[data-testid="stepper:input"]').getDOMNode());
    });
    wrapper.update();
    expect(spy).toBeCalledTimes(1);

    wrapper.setProps({ onBlur: undefined });

    act(() => {
      Simulate.blur(wrapper.find('[data-testid="stepper:input"]').getDOMNode());
    });
    wrapper.update();
    expect(spy).toBeCalledTimes(1);
  });

  it('should handle "size" prop', () => {
    const wrapper = mount(
      <Stepper />
    );

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ size: 'm' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ size: 'm' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle "onAdd/onSubtract/canAdd/canSubtract" props', () => {
    const subtractSpy = jest.fn();
    const addSpy = jest.fn();

    const wrapper = mount(
      <Stepper
        onSubtract={subtractSpy}
        onAdd={addSpy}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(subtractSpy).toHaveBeenCalledTimes(0);
    expect(addSpy).toHaveBeenCalledTimes(0);

    act(() => {
      wrapper.find('[data-testid="stepper:minus"]').first().simulate('click');
    });
    wrapper.update();

    expect(subtractSpy).toHaveBeenCalledTimes(1);
    expect(addSpy).toHaveBeenCalledTimes(0);

    act(() => {
      wrapper.find('[data-testid="stepper:plus"]').last().simulate('click');
    });
    wrapper.update();

    expect(subtractSpy).toHaveBeenCalledTimes(1);
    expect(addSpy).toHaveBeenCalledTimes(1);

    // hide buttons
    act(() => {
      wrapper.setProps({ canAdd: false, canSubtract: false });
    });
    wrapper.update();

    wrapper.find('[data-testid="stepper:minus"]').first().simulate('click');
    expect(subtractSpy).toHaveBeenCalledTimes(1);

    wrapper.find('[data-testid="stepper:plus"]').last().simulate('click');
    expect(addSpy).toHaveBeenCalledTimes(1);
  });
});
