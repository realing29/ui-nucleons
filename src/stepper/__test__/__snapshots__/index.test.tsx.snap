// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> should handle "onAdd/onSubtract/canAdd/canSubtract" props 1`] = `
<Stepper
  onAdd={[MockFunction]}
  onSubtract={[MockFunction]}
>
  <div
    className="root size-m"
    data-testid="stepper"
  >
    <ForwardRef(MinusSVG)
      className="button"
      data-testid="stepper:minus"
      onClick={[MockFunction]}
    >
      <svg
        className="button"
        data-testid="stepper:minus"
        height={16}
        onClick={[MockFunction]}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          clipRule="evenodd"
          d="M2 8a1 1 0 011-1h10a1 1 0 110 2H3a1 1 0 01-1-1z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(MinusSVG)>
    <input
      className="input"
      data-testid="stepper:input"
      onBlur={[Function]}
      onFocus={[Function]}
    />
    <ForwardRef(PlusSVG)
      className="button"
      data-testid="stepper:plus"
      onClick={[MockFunction]}
    >
      <svg
        className="button"
        data-testid="stepper:plus"
        height={16}
        onClick={[MockFunction]}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M9 3a1 1 0 00-2 0v4H3a1 1 0 000 2h4v4a1 1 0 102 0V9h4a1 1 0 100-2H9V3z"
        />
      </svg>
    </ForwardRef(PlusSVG)>
  </div>
</Stepper>
`;

exports[`<Stepper /> should handle "size" prop 1`] = `
<Stepper>
  <div
    className="root size-m"
    data-testid="stepper"
  >
    <ForwardRef(MinusSVG)
      className="button"
      data-testid="stepper:minus"
    >
      <svg
        className="button"
        data-testid="stepper:minus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          clipRule="evenodd"
          d="M2 8a1 1 0 011-1h10a1 1 0 110 2H3a1 1 0 01-1-1z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(MinusSVG)>
    <input
      className="input"
      data-testid="stepper:input"
      onBlur={[Function]}
      onFocus={[Function]}
    />
    <ForwardRef(PlusSVG)
      className="button"
      data-testid="stepper:plus"
    >
      <svg
        className="button"
        data-testid="stepper:plus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M9 3a1 1 0 00-2 0v4H3a1 1 0 000 2h4v4a1 1 0 102 0V9h4a1 1 0 100-2H9V3z"
        />
      </svg>
    </ForwardRef(PlusSVG)>
  </div>
</Stepper>
`;

exports[`<Stepper /> should handle "size" prop 2`] = `
<Stepper
  size="m"
>
  <div
    className="root size-m"
    data-testid="stepper"
  >
    <ForwardRef(MinusSVG)
      className="button"
      data-testid="stepper:minus"
    >
      <svg
        className="button"
        data-testid="stepper:minus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          clipRule="evenodd"
          d="M2 8a1 1 0 011-1h10a1 1 0 110 2H3a1 1 0 01-1-1z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(MinusSVG)>
    <input
      className="input"
      data-testid="stepper:input"
      onBlur={[Function]}
      onFocus={[Function]}
    />
    <ForwardRef(PlusSVG)
      className="button"
      data-testid="stepper:plus"
    >
      <svg
        className="button"
        data-testid="stepper:plus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M9 3a1 1 0 00-2 0v4H3a1 1 0 000 2h4v4a1 1 0 102 0V9h4a1 1 0 100-2H9V3z"
        />
      </svg>
    </ForwardRef(PlusSVG)>
  </div>
</Stepper>
`;

exports[`<Stepper /> should handle "size" prop 3`] = `
<Stepper
  size="m"
>
  <div
    className="root size-m"
    data-testid="stepper"
  >
    <ForwardRef(MinusSVG)
      className="button"
      data-testid="stepper:minus"
    >
      <svg
        className="button"
        data-testid="stepper:minus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          clipRule="evenodd"
          d="M2 8a1 1 0 011-1h10a1 1 0 110 2H3a1 1 0 01-1-1z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(MinusSVG)>
    <input
      className="input"
      data-testid="stepper:input"
      onBlur={[Function]}
      onFocus={[Function]}
    />
    <ForwardRef(PlusSVG)
      className="button"
      data-testid="stepper:plus"
    >
      <svg
        className="button"
        data-testid="stepper:plus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M9 3a1 1 0 00-2 0v4H3a1 1 0 000 2h4v4a1 1 0 102 0V9h4a1 1 0 100-2H9V3z"
        />
      </svg>
    </ForwardRef(PlusSVG)>
  </div>
</Stepper>
`;

exports[`<Stepper /> should render without props 1`] = `
<Stepper>
  <div
    className="root size-m"
    data-testid="stepper"
  >
    <ForwardRef(MinusSVG)
      className="button"
      data-testid="stepper:minus"
    >
      <svg
        className="button"
        data-testid="stepper:minus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          clipRule="evenodd"
          d="M2 8a1 1 0 011-1h10a1 1 0 110 2H3a1 1 0 01-1-1z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(MinusSVG)>
    <input
      className="input"
      data-testid="stepper:input"
      onBlur={[Function]}
      onFocus={[Function]}
    />
    <ForwardRef(PlusSVG)
      className="button"
      data-testid="stepper:plus"
    >
      <svg
        className="button"
        data-testid="stepper:plus"
        height={16}
        viewBox="0 0 16 16"
        width={16}
      >
        <path
          d="M9 3a1 1 0 00-2 0v4H3a1 1 0 000 2h4v4a1 1 0 102 0V9h4a1 1 0 100-2H9V3z"
        />
      </svg>
    </ForwardRef(PlusSVG)>
  </div>
</Stepper>
`;
